robotiq_ethercat/src/ethercat_manager.cpp

```diff
@@ -0,0 +1,226 @@
+#include <ethercat_soem/ethercatbase.h>
+#include <ethercat_soem/ethercatmain.h>
+#include <ethercat_soem/ethercatdc.h>
+#include <ethercat_soem/ethercatcoe.h>
+#include <ethercat_soem/ethercatfoe.h>
+#include <ethercat_soem/ethercatconfig.h>
+#include <ethercat_soem/ethercatprint.h>
+
+#include <ros/ros.h>
+
+namespace 
+{
+  static const unsigned THREAD_SLEEP_TIME = 10000; // 10 ms
+  static const unsigned EC_TIMEOUTMON = 500;
+
+  void handleErrors()
+  {
+    /* one ore more slaves are not responding */
+    ec_group[0].docheckstate = FALSE;
+    ec_readstate();
+    for (int slave = 1; slave <= ec_slavecount; slave++)
+    {
+      if ((ec_slave[slave].group == 0) && (ec_slave[slave].state != EC_STATE_OPERATIONAL))
+      {
+        ec_group[0].docheckstate = TRUE;
+        if (ec_slave[slave].state == (EC_STATE_SAFE_OP + EC_STATE_ERROR))
+        {
+          ROS_ERROR("ERROR : slave %d is in SAFE_OP + ERROR, attempting ack.\n", slave);
+          ec_slave[slave].state = (EC_STATE_SAFE_OP + EC_STATE_ACK);
+          ec_writestate(slave);
+        }
+        else if(ec_slave[slave].state == EC_STATE_SAFE_OP)
+        {
+          ROS_WARN("WARNING : slave %d is in SAFE_OP, change to OPERATIONAL.\n", slave);
+          ec_slave[slave].state = EC_STATE_OPERATIONAL;
+          ec_writestate(slave);
+        }
+        else if(ec_slave[slave].state > 0)
+        {
+          if (ec_reconfig_slave(slave, EC_TIMEOUTMON))
+          {
+            ec_slave[slave].islost = FALSE;
+            ROS_INFO("MESSAGE : slave %d reconfigured\n",slave);
+          }
+        }
+        else if(!ec_slave[slave].islost)
+        {
+          /* re-check state */
+          ec_statecheck(slave, EC_STATE_OPERATIONAL, EC_TIMEOUTRET);
+          if (!ec_slave[slave].state)
+          {
+            ec_slave[slave].islost = TRUE;
+            ROS_ERROR("ERROR : slave %d lost\n",slave);
+          }
+        }
+      }
+      if (ec_slave[slave].islost)
+      {
+        if(!ec_slave[slave].state)
+        {
+          if (ec_recover_slave(slave, EC_TIMEOUTMON))
+          {
+            ec_slave[slave].islost = FALSE;
+            ROS_INFO("MESSAGE : slave %d recovered\n",slave);
+          }
+        }
+        else
+        {
+          ec_slave[slave].islost = FALSE;
+          ROS_INFO("MESSAGE : slave %d found\n",slave);
+        }
+      }
+    }
+  }
+
+  void cycleWorker(boost::mutex& mutex, bool& stop_flag)
+  {
+    while (!stop_flag) 
+    {
+      int expected_wkc = (ec_group[0].outputsWKC * 2) + ec_group[0].inputsWKC;
+      int sent, wkc;
+      {
+        boost::mutex::scoped_lock lock(mutex);
+        sent = ec_send_processdata();
+        wkc = ec_receive_processdata(EC_TIMEOUTRET);
+      }
+
+      if (wkc < expected_wkc)
+      {
+        handleErrors();
+      }
+
+      usleep(THREAD_SLEEP_TIME);
+    }
+  }
+
+} // end of anonymous namespace
+
+
+robotiq_ethercat::EtherCatManager::EtherCatManager(const std::string& ifname)
```

Comment:
Anonymous name-spacing keep functions and variables private inside the compilation unit. It's equivalent to declaring everything static. I didn't want to expose those functions.